### PR TITLE
Expand doc strings with more detailed is_causal description

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Expand doc strings with more detailed is_causal description

Test Plan: sandcastle

Differential Revision: D47383610

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
-                is_causal is provides a hint that the attn_mask is the
-                causal mask.Providing incorrect hints can result in
-                incorrect execution, including forward and backward
-                compatibility.
+
+            is_causal: If ``is_causal=True``, provides a hint that ``mask``
+                is a causal mask which will enable the use of a faster fused
+                ``scaled_dot_product_attention()`` operators (SDPA) on
+                hardware where such enhanced fused SDPA are supported.
+                If ``is_causal=False`, an SDPA operator using the mask will
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5168]: nit: 
```suggestion
                If ``is_causal=False``, an SDPA operator using the mask will
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
+                be used (which may result in slower execution).  If
+                ``is_causal=None``, the TransformerEncoder and
+                TransformerDecoder modules may detect whether is_causal os
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5171]:
```suggestion
                TransformerDecoder modules may detect whether is_causal is
```

File: torch/nn/functional.py [hunk 3]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
+                a causal mask. (Detecting the presence of a causal mask
+                imposes some overhead. Thus, specifying ``is_causal=True``
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5173]: this last sentence here seems like it's missing something, would it more succinct to phrase this phrased as 
"Thus for best performance the user should specify ``is_causal``."

File: torch/nn/functional.py [hunk 4]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
+                ``is_causal=False`` when the mask is known to not be causal,
+                or when the mask should be used without attempting to detect
+                the presence of a causal mask.)
+
+                Default: ``None``; try to detect a causal mask in the input.
+
+                Warning:
+                ``is_causal`` provides a hint that ``mask`` is the
+                causal mask. Providing incorrect hints can result in
+                seemingly non-determinstic,
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5184]:
```suggestion
                seemingly non-deterministic behavior,
```

File: torch/nn/functional.py [hunk 5]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
+                including forward and backward compatibility.
+
+                Note: Individual Transformer layers in nn.Transformer,
+                nn.MultiHeadAttention, and
+                multi_had_attention_forward in nn.functional do not support
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5189]:
```suggestion
                multi_head_attention_forward in nn.functional do not support
```

File: torch/nn/functional.py [hunk 6]
```diff
@@ -5154,14 +5160,37 @@ def multi_head_attention_forward(
...
+                the detection of causal masks with ``is_causal=None``, because
+                the overhead of detecting at the layer or MHA level is
+                unlikely to yield overall performance improvement.
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:5192]:
```suggestion
                unlikely to yield an overall performance improvement.
```

File: torch/nn/modules/transformer.py [hunk 7]
```diff
@@ -378,13 +406,45 @@ def forward(self, tgt: Tensor, memory: Tensor, tgt_mask: Optional[Tensor] = None
             memory_mask: the mask for the memory sequence (optional).
             tgt_key_padding_mask: the mask for the tgt keys per batch (optional).
             memory_key_padding_mask: the mask for the memory keys per batch (optional).
-            tgt_is_causal: If specified, applies a causal mask as ``tgt mask``.
-                Default: ``None``; try to detect a causal mask.
+            tgt_is_causal: If ``tgt_is_causal=True``, provides a hint that ``mask``
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:409]:
```suggestion
            tgt_is_causal: If ``tgt_is_causal=True``, provides a hint that ``tgt_mask``
```

File: torch/nn/modules/transformer.py [hunk 8]
```diff
@@ -536,6 +598,50 @@ def forward(
...
+                ``is_causal=None``, the TransformerEncoder and
+                TransformerDecoder modules may detect whether is_causal os
+                a causal mask. (Detecting the presence of a causal mask
+                imposes some overhead. Thus, specifying ``is_causal=True``
+                when the mask is known to be a causal mask, and
+                ``is_causal=False`` when the mask is known to not be causal,
+                or when the mask should be used without attempting to detect
+                the presence of a causal mask.)
+
+                Default: ``Fa;se``.
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:616]:
```suggestion
                Default: ``False``.
```

File: torch/nn/modules/transformer.py [hunk 9]
```diff
@@ -536,6 +598,50 @@ def forward(
...
+                Note: Unlike TransformerEncoder and TransformerDecoder
+                nn.modules, ndividual Transformer layers in nn.Transformer,
+                the nn.MultiHeadAttention module, and
+                multi_had_attention_forward in nn.functional do not support
+                the detection of causal masks with ``is_causal=None``, because
+                the overhead of detecting at the layer or MHA level is
+                unlikely to yield overall performance improvement.
+
+                Warning:
+                ``is_causal`` provides a hint that ``mask`` is the
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:633]: this warning is on 618, please remove

File: torch/nn/modules/transformer.py [hunk 10]
```diff
@@ -536,6 +598,50 @@ def forward(
...
+                Default: ``Fa;se``.
+
+                Warning:
+                ``is_causal`` provides a hint that ``mask`` is the
+                causal mask. Providing incorrect hints can result in
+                seemingly non-determinstic,
+                including forward and backward compatibility.
+
+                Note: Unlike TransformerEncoder and TransformerDecoder
+                nn.modules, ndividual Transformer layers in nn.Transformer,
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:625]:
```suggestion
                nn.modules, individual Transformer layers in nn.Transformer,
```

File: torch/nn/modules/transformer.py [hunk 11]
```diff
@@ -536,6 +598,50 @@ def forward(
...
+                the nn.MultiHeadAttention module, and
+                multi_had_attention_forward in nn.functional do not support
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:627]:
```suggestion
                multi_head_attention_forward in nn.functional do not support
```

File: torch/nn/modules/transformer.py [hunk 12]
```diff
@@ -754,13 +860,47 @@ def forward(
...
+                the overhead of detecting at the layer or MHA level is
+                unlikely to yield overall performance improvement.
+
+                Warning:
+                ``tgt_is_causal`` provides a hint that ``tgt_mask`` is the
+                causal mask. Providing incorrect hints can result in
+                seemingly non-determinstic,
+                including forward and backward compatibility.
+
+                Note: Unlike TransformerEncoder and TransformerDecoder in
```
Comment (mikaylagawarecki) [inline torch/nn/modules/transformer.py:896]: Duplicated

**Review by mikaylagawarecki — COMMENTED** [review]
As discussed offline, I only reviewed the docstrings and not the code from the PR stacked below.

General comments here -- the content makes sense (albeit a bit long). However, it seems like there is a lot of replication. I didn't replicate comments from all the docstrings to the others since they have roughly the same content. 
1) For ease of maintainability it would be nice if we could keep the long docstring for `is_causal` in one place and link to it from the other places where there is `is_causal`/`{*}_is_causal`. 
 *If it would be useful, I can look into whether this is possible*
2) It would be nice to use sphinx features like `:func:` or `:meth:` when we refer to things `e.g. nn.Transformer`, `nn.MultiheadAttention` and double backticks on all argument names.  I neglected to make individual comments for all these occurrences. 

3) `..note::` and `..warning::` are also a nice-to-have but I suppose the omission of these were intentional since we want to inline these notes and warniings in the docstring for `is_causal`.

**Comment by mikekgfb** [comment]
> 1. For ease of maintainability it would be nice if we could keep the long docstring for `is_causal` in one place and link to it from the other places where there is `is_causal`/`{*}_is_causal`.
>    _If it would be useful, I can look into whether this is possible_

I don't know what our practice is.  For sure, maintaining the replicas is a pain. OTOH, I wonder if I look at TransformerEncoder, would I click on the TransformerDecoder link?    There just doesn't seem to be a place that can consistently apply to multiple functions, and be found by users.

**Comment by mikaylagawarecki** [comment]
@mikekgfb That is a fair point. You are right that the convention does seem to be to replicate the documentation when it is rendered! That said at least from a code perspective perhaps we could have [something like this](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L2031) (e.g. `is_causal_note`) so that the note itself only occurs once in the codebase. What do you think of this?

**Pull request closed without merging**